# 05. Testes de Hipóteses
---

<img src="https://selecao.letscode.com.br/favicon.png" width="40px" style="position: absolute; top: 20px; right: 20px; border-radius: 5px;" />

Testes estatísticos são regras de decisão que permitem avaliar a razoabilidade das hipóteses feitas sobre os parâmetros populacionais e aceitá-las ou rejeitá-las como provavelmente verdadeiras ou falsas tendo como base uma amostra.

Nos testes de hipóteses, define uma regra de decisão para rejeitar uma hipótese estatística com base nos elementos amostrais.

Para testar um parâmetro populacional, devemos afirmar, cuidadosamente, um par de hipóteses: uma que represente a afirmação e outra que represente o seu complemento. Quando uma dessas hipóteses for falsa, a outra deverá ser verdadeira. Essas duas hipóteses são chamadas de hipótese nula e hipótese alternativa.

**Importação das Bibliotecas**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import norm, t as t_student

sns.set()

## <font color="blue">1. Teste de Normalidade</font>
***

**Carregando nosso dataset**

### Importando a biblioteca
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.normaltest.html

A função *normaltest* testa a hipótese nula $H_0$ de que a amostra é proveniente de uma distribuição normal.

- $H_0$: hipótese nula  
- $H_1$: hipótese alternativa

### Definindo a significância do teste ($\alpha$)

### Testando a variável Renda

In [3]:
# estatística de teste
# p-valor (p-value)


### <font color='red'>Critério do valor $p$</font>

> ### Rejeitar $H_0$ se o valor $p\leq \alpha \, (0,05)$

In [4]:
# H0: a amostra segue uma distribuição normal
# H1: a amostra NÃO segue uma distribuição normal

### Testando a variável Altura

### <font color='red'>Critério do valor $p$</font>

> ### Rejeitar $H_0$ se o valor $p\leq 0,05$

## <font color="blue">2. Etapas Básicas de um Teste</font>
***

### **Passo 1** - Formulação das hipóteses $H_0$ e $H_1$

> ### <font color='red'>Pontos importantes</font>
> - A hipótese nula sempre afirma uma igualdade ou propriedade populacional, e $H_1$ a desigualdade que nega $H_0$.
> - No caso da hipótese nula $H_0$ a igualdade pode ser representada por uma igualdade simples "$=$" ou por "$\geq$" e "$\leq$". Sempre complementar ao estabelecido pela hipótese alternativa.
> - A hipótese alternativa $H_1$ deve definir uma desigualdade que pode ser uma diferença simples "$\neq$" ou dos tipos "$>$" e "$<$".


### **Passo 2** - Escolha da distribuição amostral adequada

> ### <font color='red'>Pontos importantes</font>
> - Quando o tamanho da amostra tiver 30 elementos ou mais, deve-se utilizar a distribuição normal, como estabelecido pelo **teorema do limite central**.
> - Para um tamanho de amostra menor que 30 elementos, e se pudermos afirmar que a população se distribui aproximadamente como uma normal e o desvio padrão populacional for conhecido, deve-se utilizar a distribuição normal.
> - Para um tamanho de amostra menor que 30 elementos, e se pudermos afirmar que a população se distribui aproximadamente como uma normal e o desvio padrão populacional não for desconhecido, deve-se utilizar a distribuição t de Student.

<img style="display: block; margin: 25px auto" src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img003.png' width=70%>

### **Passo 3** - Fixação da significância do teste ($\alpha$), que define as regiões de aceitação e rejeição das hipóteses (os valores mais freqüentes são 10%, 5% e 1%);

> ### <font color='red'>Pontos importantes</font>
> - O **nível de confiança** ($1 - \alpha$) representa a probabilidade de acerto da estimativa. De forma complementar o **nível de significância** ($\alpha$) expressa a probabilidade de erro da estimativa.
>
> <img alt="Níveis de Confiança e significância" src="https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img001.png" style="display: block; margin: 25px auto" />
>
> - O **nível de confiança** representa o grau de confiabilidade do resultado da estimativa estar dentro de determinado intervalo. Quando fixamos em uma pesquisa um **nível de confiança** de 95%, por exemplo, estamos assumindo que existe uma probabilidade de 95% dos resultados da pesquisa representarem bem a realidade, ou seja, estarem corretos.
>
> <img alt="Áreas de Aceitação e Rejeição" src="https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img002.png" style="display: block; margin: 25px auto; border-radius: 10px" />

### **Passo 4** - cálculo da estatística-teste e verificação desse valor com as áreas de aceitação e rejeição do teste;

> ### <font color='red'>Pontos importantes</font>
> - Nos testes paramétricos, distância relativa entre a estatística amostral e o valor alegado como provável.
> - Neste passo são obtidas as estatísticas amostrais necessárias à execução do teste (média, desvio-padrão, graus de liberdade etc.)


### **Passo 5** - Aceitação ou rejeição da hipótese nula.

> ### <font color='red'>Pontos importantes</font>
> - No caso de o intervalo de aceitação conter a estatística-teste, aceita-se $H_0$ como estatisticamente válido e rejeita-se $H_1$ como tal.
> - No caso de o intervalo de aceitação não conter a estatística-teste, rejeita-se $H_0$ e aceita-se $H_1$ como provavelmente verdadeira. 
> - A aceitação também se verifica com a probabilidade de cauda (p-valor): se maior que $\alpha$, aceita-se $H_0$.

---

## <font color=blue>3. TESTES PARAMÉTRICOS</font>
***

Quando um teste assume determinadas premissas sobre como os parâmetros de uma população se distribuem, estamos trabalhando com **Testes Paramétricos**.

### <font color=blue>3.1 Teste Bicaudal</font>
***

#### <font color='red'>Problema</font>

A empresa **Suco Bom** produz **sucos de frutas em embalagens de 500 ml**. Seu processo de produção é quase todo automatizado e as embalagens de sucos são preenchidas por uma máquina que às vezes apresenta um certo desajuste, levando a erros no preenchimento das embalagens para mais ou menos conteúdo. Quando o volume médio cai abaixo de 500 ml, a empresa se preocupa em perder vendas e ter problemas com os orgãos fiscalizadores. Quando o volume passa de 500 ml, a empresa começa a se preocupar com prejuízos no processo de produção.

O setor de controle de qualidade da empresa **Suco Bom** extrai, periodicamente, **amostras de 50 embalagens** para monitorar o processo de produção. Para cada amostra, é realizado um **teste de hipóteses** para avaliar se o maquinário se desajustou. A equipe de controle de qualidade assume um **nível de significância de 5%**.

Suponha agora que uma **amostra de 50 embalagens** foi selecionada e que a **média amostral observada foi de 503,24 ml**. **Esse valor de média amostral é suficientemente maior que 500 ml para nos fazer rejeitar a hipótese de que a média do processo é de 500 ml, ao nível de significância de 5%?**

---

O **teste bicaudal** é muito utilizado em **testes de qualidade**, como o apresentado em nosso problema acima. Outro exemplo é a avaliação de peças que devem ter um encaixe perfeito (porcas e parafusos, chaves e fechaduras).

![Teste Bicaudal](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img005.png)

---

In [12]:
u0 = 500
media_amostral = 485
desvio_amostral = 20
n = 30

In [6]:
sig = 0.05
conf = 1 - sig

In [7]:
z_alpha_2 = norm.ppf(conf + sig / 2)
z_alpha_2

1.959963984540054

In [17]:
z = (media_amostral - u0) / (desvio_amostral / np.sqrt(n))
z

-4.107919181288746

In [20]:
if z <= -z_alpha_2 or z >= z_alpha_2:
    print('Reijeita H0')
else:
    print('Aceita H0')

Reijeita H0


In [19]:
p_value = norm.sf(z) * 2
p_value

1.999960076025231

A empresa Limpa Esgoto garante ser capaz de realizar o tratamento de esgoto e obter, no máximo, 150 g de impurezas para cada mil litros de esgoto tratado. Vinte amostras de mil litros de esgoto apresentaram, em média, 230 g de impurezas e desvio padrão amostral igual a 90 g.  

Assumindo alfa igual a 5% e população normalmente distribuída, seria possível discordar da empresa Limpa Esgoto? Assinale a alternativa que apresenta a estatística de teste e a decisão correta do teste.  

In [37]:
from scipy.stats import t as t_student

In [45]:
# H0 >= 350  H1 < 350
u0 = 350
n = 35
media_amostral = 330
desvio_amostral = 80
sig = 0.05
gl = n -1

In [46]:
conf = 1 - sig
z_alpha_2 = norm.ppf(conf + sig / 2)
print(f'Z_alpha = {z_alpha_2}')
t_alpha = t_student.ppf(conf, df=gl)
print(f'T_alpha = {t_alpha}')

t = (media_amostral - u0) / (desvio_amostral / np.sqrt(n))
print(f'T = {t}')

if t >= t_alpha:
    print('Reijeita H0')
else:
    print('Aceita H0')

p_value = norm.sf(t)
print(f'P value = {p_value}')

Z_alpha = 1.959963984540054
T_alpha = 1.6909242507706543
T = -1.479019945774904
Aceita H0
P value = 0.9304325079659079


Um fabricante de cosméticos afirma que a adição de um novo composto químico em sua linha de shampoos consegue promover em mais de 2 centímetros o crescimento dos fios de cabelo em um período de 60 dias. Duas amostras de pessoas foram selecionadas e testadas, uma utilizando o shampoo novo (com o composto) e a outra com o shampoo antigo (sem o composto).  

Os resultados (crescimento dos fios de cabelo em centímetros) podem ser verificados na tabela abaixo:  

Um fabricante de cosméticos afirma que a adição de um novo composto químico em sua linha de shampoos consegue promover em mais de 2 centímetros o crescimento dos fios de cabelo em um período de 60 dias. Duas amostras de pessoas foram selecionadas e testadas, uma utilizando o shampoo novo (com o composto) e a outra com o shampoo antigo (sem o composto).  

Os resultados (crescimento dos fios de cabelo em centímetros) podem ser verificados na tabela abaixo:  

---

#### Dados do problema

In [5]:
amostra = [509, 505, 495, 510, 496, 509, 497, 502, 503, 505, 
           501, 505, 510, 505, 504, 497, 506, 506, 508, 505, 
           497, 504, 500, 498, 506, 496, 508, 497, 503, 501, 
           503, 506, 499, 498, 509, 507, 503, 499, 509, 495, 
           502, 505, 504, 509, 508, 501, 505, 497, 508, 507]

In [6]:
# Passar para um dataframe
dados = pd.DataFrame({ 'Volume': amostra })

In [7]:
dados.head()

Volume
0     509
1     505
2     495
3     510
4     496

In [8]:
dados.mean()

Volume    503.24
dtype: float64

In [9]:
# Média da amostra
media_amostral = dados.Volume.mean()

media_amostral

503.24

In [10]:
# Desvio padrão (amostral)
desvio_amostral = dados.Volume.std()
desvio_amostral

4.483803050527348

In [11]:
# definindo os demais dados do problema
u0 = 500
n = len(dados)
n

50

#### **Passo 1** - formulação das hipóteses $H_0$ e $H_1$

<font color='red'>Lembre-se, a hipótese nula sempre contém a alegação de igualdade</font>

$H_0: \mu = 500 \, ml$

$H_1: \mu \neq 500 \, ml$

---

#### **Passo 2** - escolha da distribuição amostral adequada

<img style="border-radius: 10px" src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img003.png' width=50%>

O tamanho da amostra é maior que 30?
Resp.: Sim

O desvio padrão populacional é conhecido?
Resp.: Não

---

#### **Passo 3** - fixação da significância do teste ($\alpha$)

In [12]:
significancia = 0.05
confianca = 0.95

#### Obtendo $z_{\alpha/2}$

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.norm.html

In [13]:
z_alpha_2 = norm.ppf(confianca + significancia / 2)
z_alpha_2

1.959963984540054

<img style="border-radius: 10px" alt="Região de Aceitação" src="https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img006.png" />

---

#### **Passo 4** - cálculo da estatística-teste e verificação desse valor com as áreas de aceitação e rejeição do teste

$$z = \frac{\bar{x} - \mu_0}{\frac{s}{\sqrt{n}}}$$

In [14]:
z = (media_amostral - u0) / (desvio_amostral / np.sqrt(n))
z # Estatística de teste

5.109559775991876

![Estatística-Teste](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img007.png)

---

#### **Passo 5** - Aceitação ou rejeição da hipótese nula

<img style="border-radius: 10px" src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img013.png' width=80%>

#### <font color='green'>Critério do valor crítico</font>

> Teste Bicaudal
> Rejeitar $H_0$ se $z \leq -z_{\alpha / 2}$ ou se $z \geq z_{\alpha / 2}$

In [15]:
z <= -z_alpha_2 or z >= z_alpha_2

True

<font color='orange'>Conclusão: Como a média amostral $\bar{x}$ é significativamente maior que 500 ml, rejeitamos $H_0$. Neste caso, devem ser tomadas providências para ajustar o maquinário que preenche as embalagens.</font>

#### <font color='green'>Critério do $p-valor$</font>

> Teste Bicaudal  
> Rejeitar $H_0$ se o valor $p\leq\alpha$

In [16]:
p_value = norm.sf(5.11) * 2
p_value

3.221588122442738e-07

In [17]:
if p_value <= significancia:
    print('Rejeita o H0, ou seja, a hipótese nula => A média do volume das embalagens de suco NÃO é igual a 500 ml')
else:
    print('Aceita o H0, ou seja, a hipótese nula => A média do volume das embalagens de suco É igual a 500 ml')

Rejeita o H0, ou seja, a hipótese nula => A média do volume das embalagens de suco NÃO é igual a 500 ml


https://www.statsmodels.org/stable/generated/statsmodels.stats.weightstats.ztest.html

In [36]:
from statsmodels.stats.weightstats import ztest

In [20]:
dados.head()

Volume
0     509
1     505
2     495
3     510
4     496

In [21]:
ztest(x1=dados['Volume'], value=500, alternative='two-sided')

(5.109559775991873, 3.229103172445771e-07)

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.DescrStatsW.html

In [22]:
from statsmodels.stats.weightstats import DescrStatsW

In [23]:
teste = DescrStatsW(dados['Volume'])

In [24]:
teste.ztest_mean(value=500)

(5.109559775991874, 3.2291031724457596e-07)

---

### <font color=green>3.2 Teste Unicaudal</font>
***

#### <font color='red'>Problema</font>

Um famoso fabricante de refrigerantes alega que uma lata de 350 ml de seu principal produto contém, **no máximo**, **37 gramas de açúcar**. Esta alegação nos leva a entender que a quantidade média de açúcar em uma lata de refrigerante deve ser **igual ou menor que 37 g**.

Um consumidor desconfiado e com conhecimentos em inferência estatística resolve testar a alegação do fabricante e seleciona, aleatóriamente, em um conjunto de estabelecimentos distintos, **uma amostra de 25 latas** do refrigerante em questão. Utilizando o equipamento correto o consumidor obteve as quantidades de açúcar em todas as 25 latas de sua amostra. 

**Assumindo que essa população se distribua aproximadamente como uma normal e considerando um nível de significância de 5%, é possível aceitar como válida a alegação do fabricante?**

#### 3.2.1. Conhecendo a Distribuição $t$-student
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.t.html

<img src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img004.png' width='250px'>

**Propriedades**

- A função densidade da distribuição t de Student tem a mesma forma em sino da distribuição Normal, mas reflete a maior variabilidade (com curvas mais alargadas) que é de se esperar em amostras pequenas.
- Quanto maior o grau de liberdade, mais a distribuição t de Student se aproxima da distribuição Normal.

---

#### 3.2.2. Teste Unicaudal

Os **testes unicaudais** verificam as variáveis em relação a um piso ou a um teto e avaliam os valores máximos ou mínimos esperados para os parâmetros em estudo e a chance de as estatísticas amostrais serem inferiores ou superiores a dado limite.

<img style="border-radius: 10px" src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img008.png' width='700px'>

##### Dados do problema

In [25]:
amostra = [37.27, 36.42, 34.84, 34.60, 37.49, 
           36.53, 35.49, 36.90, 34.52, 37.30, 
           34.99, 36.55, 36.29, 36.06, 37.42, 
           34.47, 36.70, 35.86, 36.80, 36.92, 
           37.04, 36.39, 37.32, 36.64, 35.45]

In [26]:
len(amostra)

25

In [27]:
# Converter para um DataFrame
dados = pd.DataFrame({ 'gramas': amostra })

In [28]:
# Visualização
dados.head()

gramas
0   37.27
1   36.42
2   34.84
3   34.60
4   37.49

In [29]:
# media amostral
media_amostral = dados.gramas.mean()
media_amostral

36.250400000000006

In [30]:
# desvio padrão amostral
desvio_amostral = dados.gramas.std()
desvio_amostral

0.9667535018469453

In [31]:
# número de amostras e número de graus de liberdade
n = len(dados)
gl = n - 1

#### **Passo 1** - formulação das hipóteses $H_0$ e $H_1$

<font color='red'>Lembre-se, a hipótese nula sempre contém a alegação de igualdade</font>

$H_0: \mu \leq 37 \, g$

$H_1: \mu > 37 \, g$

---

#### **Passo 2** - escolha da distribuição amostral adequada

<img style="border-radius: 10px" src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img003.png' width=60%>

O tamanho da amostra é maior que 30?
Resp.: Não

Podemos afirmar que a população se distribui aproximadamente como uma normal?
Resp.: Sim

O desvio padrão populacional é conhecido?
Resp.: Não

---

#### **Passo 3** - Fixação da significância do teste ($\alpha$)

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.t.html

In [32]:
significancia = 0.05
confianca = 1 - significancia

##### Obtendo $t_{\alpha}$

In [33]:
from scipy.stats import t as t_student

In [34]:
t_alpha = t_student.ppf(0.95, df=gl)
t_alpha

1.7108820799094275

![Região de Aceitação](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img009.png)

---

#### **Passo 4** - cálculo da estatística-teste e verificação desse valor com as áreas de aceitação e rejeição do teste

$$t = \frac{\bar{x} - \mu_0}{\frac{s}{\sqrt{n}}}$$

In [35]:
t = (media_amostral - 37) / (desvio_amostral / np.sqrt(n))
t

-3.876893119952045

![Estatística-Teste](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img010.png)

In [36]:
# Aceito a hipótese a nula

---

#### **Passo 5** - Aceitação ou rejeição da hipótese nula

<img style="border-radius: 10px" src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img013.png' width=70%>

#### <font color='red'>Critério do valor crítico</font>

> Teste Unicaudal Superior  
> Rejeitar $H_0$ se $t \geq t_{\alpha}$

In [37]:
t > t_alpha

False

<font color='green'>Conclusão: Com um nível de confiança de 95% não podemos rejeitar $H_0$, ou seja, a alegação do fabricante é verdadeira.</font>

#### <font color='red'>Critério do $p$ valor</font>

> Teste Unicaudal Superior  
> Rejeitar $H_0$ se o valor $p\leq\alpha$

In [38]:
p_value = t_student.sf(t, df=gl)

In [39]:
p_value

0.999640617030382

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.DescrStatsW.html

In [40]:
dados.gramas.size

25

In [41]:
teste = DescrStatsW(dados.gramas)

In [42]:
# ztest_mean: Distribuição Normal
# ttest_mean: Distribuição t-Student

In [43]:
teste.ttest_mean(value=37, alternative='larger')

(-3.87689311995208, 0.9996406170303819, 24.0)

---

### <font color=green>3.4 Testes para Duas Amostras</font>
***

#### <font color='red'>**Problema**</font>

Em nosso dataset temos os rendimento dos chefes de domicílio obtidos da Pesquisa Nacional por Amostra de Domicílios - PNAD no ano de 2015. Um problema bastante conhecido em nosso país diz respeito a desigualdade de renda, principalmente entre homens e mulheres.

Duas amostras aleatórias, uma de **500 homens** e outra com **500 mulheres**, foram selecionadas em nosso dataset. Com o objetivo de comprovar tal desigualdade, **teste a igualdade das médias** entre estas duas amostras com um nível de **significância de 1%**.

---

É possível também utilizar testes de hipóteses para comparar duas diferentes amostras. Neste tipo de teste se deseja decidir se uma amostra é diferente da outra.

#### Seleção das amostras

In [2]:
df = pd.read_csv('https://s3-sa-east-1.amazonaws.com/lcpi/360d2eaa-970a-4204-8004-53bb42af101a.csv')

In [3]:
df.head()

UF  Sexo  Idade  Cor  Anos de Estudo  Renda    Altura
0  11     0     23    8              12    800  1.603808
1  11     1     23    2              12   1150  1.739790
2  11     1     35    8              15    880  1.760444
3  11     0     46    2               6   3500  1.783158
4  11     1     47    8               9    150  1.690631

In [28]:
df.shape

(76840, 7)

In [4]:
homens = df[df['Sexo'] == 0].sample(n=500, random_state=101)

In [5]:
mulheres = df.query('Sexo == 1').sample(n=500, random_state=101)

In [6]:
homens.shape

(500, 7)

In [7]:
mulheres.shape

(500, 7)

#### Obtendo dados do problema

- Média e desvio padrão das mulheres
- Média e desvio padrão dos homens
- Número de amostras (homens e mulheres)
- Nível de significância e confiança

In [8]:
media_H = homens.Renda.mean()
desvio_H = homens.Renda.std()

media_H, desvio_H

(2142.608, 2548.050802499875)

In [9]:
media_M = mulheres.Renda.mean()
desvio_M = mulheres.Renda.std()

media_M, desvio_M

(1357.528, 1569.9011907484578)

In [10]:
nH = len(homens)
nM = len(mulheres)
D0 = 0

In [11]:
significancia = 0.01

In [13]:
confianca = 1 - significancia

---

#### **Passo 1** - formulação das hipóteses $H_0$ e $H_1$

<font color='red'>Lembre-se, a hipótese nula sempre contém a alegação de igualdade</font>

$\mu_1 \Rightarrow$ Média das rendas dos chefes de domicílios do sexo masculino  
$\mu_2 \Rightarrow$ Média das rendas dos chefes de domicílios do sexo feminino  

\begin{cases}
H_0: \mu_1 \leq \mu_2\\
H_1: \mu_1 > \mu_2
\end{cases}

\begin{cases}
H_0: \mu_1 -\mu_2 \leq 0\\
H_1: \mu_1 -\mu_2 > 0
\end{cases}

In [62]:
sn = np.array([3.4,4.9,2.8,5.5,3.7,2.5,4.3,4.6,3.7,3.4])
sa = np.array([0.3,1.2,1.2,1.7,1.1,0.6,1.2,1.5,0.5,0.7])
len(sn) + len(sa) >= 30
media_sn = sn.mean()
media_sa = sa.mean()
desvio_sn = sn.std()
desvio_sa = sa.std()
gl = len(sn) + len(sa) - 2
t_alpha = t_student.ppf(0.95, df=gl)
t_alpha

1.7340636066175354

In [60]:
numerador = (media_sn - media_sa) - 0
denominador = np.sqrt((desvio_sn **2 / len(sn)) + (desvio_sa**2 / len(sa)))
t = numerador / denominador
t

9.192321660283243

In [61]:
t > t_alpha

True

#### **Passo 2** - escolha da distribuição amostral adequada

<img style="border-radius: 10px" src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img003.png' width=60%>

<font  color='red'><strong>Observação importante</strong></font>

> **Em testes que envolvam duas amostras com o emprego da tabela $t$ de Student, o número de graus de liberdade será sempre igual a $n_1 + n_2 - 2$**

O tamanho da amostra é maior que 30?
Resp.: Sim

O desvio padrão populacional é conhecido?
Resp.: Não

---

#### **Passo 3** - fixação da significância do teste ($\alpha$)

#### Diferenças entre `ppf`, `cdf` e `sf`

1. `ppf`: **passamos um valor de probabilidade** e ele **retorna um valor no eixo** `z` (norm) ou `t` (t_student).
    - Quando estamos com uma Normal/t-Student bicaudal, então, passamos $\beta + \alpha / 2$
    - Quando estamos com uma Normal/t-Student unicaudal à esquerda (confiança), então, passamos somente o $\beta$
    - Quando estamos com uma Normal/t-Student unicaudal à direita (confiança), então, passamos somente o $\alpha$
2. `cdf`: **passamos um valor no eixo** `z` (norm) ou `t` (t_student) e ele **retorna um valor de probabilidade**, à esqueda do valor passado como parâmetro.
3. `sf`: **passamos um valor no eixo** `z` (norm) ou `t` (t_student) e ele **retorna um valor de probabilidade**, à direita do valor passado como parâmetro.

Legenda
- $\beta$: confiança
- $\alpha$: significância

In [55]:
z_alpha = norm.ppf(confianca)
z_alpha

2.3263478740408408

![Região de Aceitação](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img011.png)

---

#### **Passo 4** - cálculo da estatística-teste e verificação desse valor com as áreas de aceitação e rejeição do teste

$$z = \frac{(\bar{x_1} - \bar{x_2})-D_0}{\sqrt{\frac{s_1^2}{n_1} + \frac{s_2^2}{n_2}}}$$

In [14]:
numerador = (media_H - media_M) - D0
denominador = np.sqrt((desvio_H **2 / nH) + (desvio_M**2 / nM))
z = numerador / denominador
z

5.86562005776475

![Estatística-Teste](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img012.png)

---

#### **Passo 5** - Aceitação ou rejeição da hipótese nula

<img style="border-radius: 10px" src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img014.png' width=70%>

#### <font color='red'>Critério do valor crítico</font>

> Teste Unicaudal
> Rejeitar $H_0$ se $z \geq z_{\alpha}$

<font color='green'>Conclusão: Com um nível de confiança de 99% rejeitamos $H_0$, isto é, concluímos que a média das rendas dos chefes de domicílios do sexo masculino é maior que a média das rendas das chefes de domicílios do sexo feminino. Confirmando a alegação de desigualdade de renda entre os sexos.</font>

#### <font color='red'>Critério do valor $p$</font>

> Teste Unicaudal
> Rejeitar $H_0$ se o valor $p\leq\alpha$

In [21]:
p_value = norm.sf(z)

In [22]:
p_value

2.2372867859458743e-09

In [20]:
p_value <= 0.01 # Rejeita H0

True

### Facilitando os cálculos com o Python
---

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.DescrStatsW.html

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.CompareMeans.html

In [61]:
from statsmodels.stats.weightstats import DescrStatsW, CompareMeans

**Utilizando o CompareMeans**

In [62]:
teste_homens = DescrStatsW(homens.Renda)

In [63]:
teste_mulheres = DescrStatsW(mulheres.Renda)

In [64]:
teste = CompareMeans(teste_homens, teste_mulheres)

In [65]:
teste.ztest_ind(alternative='larger', value=D0)

(5.865620057764754, 2.2372867859458255e-09)

---

## <font color=blue>4. TESTES NÃO PARAMÉTRICOS</font>
***

O trabalho com pequenas amostras pode levar a não aceitação da validade do teorema central do limite e também na impossibilidade de fazer suposições sobre a distribuição da variável avaliada. Quando isso ocorre torna-se necessária a aplicação de testes não paramétricos. Nos testes não paramétricos, não fazemos hipóteses sobre a distribuição (de probabilidade) das quais as observações são extraídas.

### <font color='red'>Problema</font>

Antes de cada partida do campeonato nacional de futebol, as moedas utilizadas pelos árbitros devem ser verificadas para se ter certeza de que não são viciadas, ou seja, que não tendam para determinado resultado. Para isso um teste simples deve ser realizado antes de cada partida. Este teste consiste em lançar a moeda do jogo **50 vezes** e contar as frequências de **CARAS** e **COROAS** obtidas. A tabela abaixo mostra o resultado obtido no experimento:

||CARA|COROA|
|-|-|-|
|Observado|17|33|
|Esperado|25|25|

A um **nível de significância de 5%**, é possível afirmar que a moeda não é honesta, isto é, que a moeda apresenta uma probabilidade maior de cair com a face **CARA** voltada para cima?

### <font color=DodgerBlue>4.1 Teste do Qui-Quadrado ( $\chi^2$)</font>
***

Também conhecido como teste de adequação ao ajustamento, seu nome se deve ao fato de utilizar uma variável estatística padronizada, representada pela letra grega qui ( $\chi$) elevada ao quadrado. A tabela com os valores padronizados e como obtê-la podem ser vistos logo abaixo.

O teste do $\chi^2$ testa a hipótese nula de não haver diferença entre as frequências observadas de um determinado evento e as frequências que são realmente esperadas para este evento.

Os passos de aplicação do teste são bem parecidos aos vistos para os testes paramétricos.

![Região de Aceitação](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img017.png)

<img src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img016.png' width='250px'>

Tabela com os valores de $\chi_p^2$ em função dos graus de liberdade $(n - 1)$ e de $p = P(\chi^2 \leq \chi_p^2)$

#### <font color='red'>Problema</font>

Antes de cada partida do campeonato nacional de futebol, as moedas utilizadas pelos árbitros devem ser verificadas para se ter certeza de que não são viciadas, ou seja, que não tendam para determinado resultado. Para isso um teste simples deve ser realizado antes de cada partida. Este teste consiste em lançar a moeda do jogo **50 vezes** e contar as frequências de **CARAS** e **COROAS** obtidas. A tabela abaixo mostra o resultado obtido no experimento:

||CARA|COROA|
|-|-|-|
|Observado|17|33|
|Esperado|25|25|

A um **nível de significância de 5%**, é possível afirmar que a moeda não é honesta, isto é, que a moeda apresenta uma probabilidade maior de cair com a face **CARA** voltada para cima?

---

#### Dados do problema

In [66]:
F_observada = [17, 33]
F_esperada = [25, 25]

significancia = 0.05
confianca = 1 - significancia

k = 2 # Número de eventos possíveis
graus_de_liberdade = k - 1

#### **Passo 1** - formulação das hipóteses $H_0$ e $H_1$

<font color='red'>Lembre-se, a hipótese nula sempre contém a alegação de igualdade</font>

$H_0: F_{CARA} = F_{COROA}$

$H_1: F_{CARA} \neq F_{COROA}$

---

#### **Passo 2** - fixação da significância do teste ($\alpha$)

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi.html

In [99]:
significancia

0.050000000000000044

In [100]:
confianca

0.95

#### Obtendo $\chi_{\alpha}^2$

In [110]:
from scipy.stats import chi2

In [105]:
chi2_alpha = chi2.ppf(confianca, df=graus_de_liberdade)

In [106]:
chi2_alpha 

3.841458820694124

![Região de Aceitação](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img018.png)

---

#### **Passo 3** - cálculo da estatística-teste e verificação desse valor com as áreas de aceitação e rejeição do teste

$$\chi^2 = \sum_{i=1}^{k}{\frac{(F_{i}^{Obs} - F_{i}^{Esp})^2}{F_{i}^{Esp}}}$$

Onde

$F_{i}^{Obs}$ = frequência observada para o evento $i$

$F_{i}^{Esp}$ = frequência esperada para o evento $i$

$k$ = total de eventos possíveis

In [109]:
chi2_teste = 0

for i in range(k):
    chi2_teste += (F_observada[i] - F_esperada[i])**2 / F_esperada[i]

chi2_teste

5.12

![Estatística-Teste](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img019.png)

---

#### **Passo 4** - Aceitação ou rejeição da hipótese nula

<img src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img020.png' width=80%>

#### <font color='red'>Critério do valor crítico</font>

> Rejeitar $H_0$ se $\chi_{teste}^2 > \chi_{\alpha}^2$

In [112]:
chi2_teste > chi2_alpha

True

<font color='green'>Conclusão: Com um nível de confiança de 95% rejeitamos a hipótese nula ($H_0$) e concluímos que as frequências observadas e esperadas são discrepantes, ou seja, a moeda não é honesta e precisa ser substituída.</font>

#### <font color='red'>Critério do valor $p$</font>

>Rejeitar $H_0$ se o valor $p\leq\alpha$

In [113]:
p_value = chi2.sf(chi2_teste, df=graus_de_liberdade)

In [114]:
p_value

0.023651616655356

In [115]:
p_value < significancia

True

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chisquare.html

In [116]:
from scipy.stats import chisquare

In [117]:
chisquare(F_observada), F_esperada

(Power_divergenceResult(statistic=5.12, pvalue=0.023651616655356), [25, 25])

---

### <font color=green>4.2 Teste Wilcoxon</font>
Comparação de duas populações - amostras dependentes
***

Empregado quando se deseja comparar duas amostras relacionadas, amostras emparelhadas. Pode ser aplicado quando se deseja testar a diferença de duas condições, isto é, quando um mesmo elemento é submetido a duas medidas.

#### <font color='red'>Problema</font>

Um novo tratamento para acabar com o hábito de fumar está sendo empregado em um grupo de **35 pacientes** voluntários. De cada paciente testado foram obtidas as informações de quantidades de cigarros consumidos por dia antes e depois do término do tratamento. Assumindo um **nível de confiança de 95%** é possível concluir que, depois da aplicação do novo tratamento, houve uma mudança no hábito de fumar do grupo de pacientes testado?

#### Dados do problema

In [118]:
fumo = {
    'Antes': [39, 25, 24, 50, 13, 52, 21, 29, 10, 22, 50, 15, 36, 39, 52, 48, 24, 15, 40, 41, 17, 12, 21, 49, 14, 55, 46, 22, 28, 23, 37, 17, 31, 49, 49],
    'Depois': [16, 8, 12, 0, 14, 16, 13, 12, 19, 17, 17, 2, 15, 10, 20, 13, 0, 4, 16, 18, 16, 16, 9, 9, 18, 4, 17, 0, 11, 14, 0, 19, 2, 9, 6]
}

In [119]:
confianca = 0.95
significancia = 1 - confianca
n = 35

In [120]:
fumo = pd.DataFrame(fumo)
fumo.head()

Antes  Depois
0     39      16
1     25       8
2     24      12
3     50       0
4     13      14

In [70]:
from scipy.stats import normaltest

In [121]:
normaltest(fumo.Antes)

NormaltestResult(statistic=15.652924739801158, pvalue=0.0003990346232004849)

In [71]:
normaltest(fumo.Depois)

NormaltestResult(statistic=5.229080349087136, pvalue=0.07320144084427976)

In [72]:
media_antes = fumo.Antes.mean()
media_antes

31.857142857142858

In [73]:
media_depois = fumo.Depois.mean()
media_depois

11.2

#### **Passo 1** - formulação das hipóteses $H_0$ e $H_1$

<font color='red'>Lembre-se, a hipótese nula sempre contém a alegação de igualdade</font>

$H_0: \mu_{antes} = \mu_{depois}$

$H_1: \mu_{antes} \neq \mu_{depois}$

---

#### **Passo 2** - escolha da distribuição amostral adequada

O tamanho da amostra é maior que 20?
Resp.: Sim

---

#### **Passo 3** - fixação da significância do teste ($\alpha$)

#### Obtendo $z_{\alpha/2}$

In [74]:
z_alpha_2 = norm.ppf(confianca + significancia / 2)

In [75]:
z_alpha_2

1.959963984540054

![Região de Aceitação](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img006.png)

---

#### **Passo 4** - cálculo da estatística-teste e verificação desse valor com as áreas de aceitação e rejeição do teste

$$Z = \frac{T - \mu_T}{\sigma_T}$$

Onde

$T$ = menor das somas de postos de mesmo sinal

$$\mu_T = \frac{n(n+1)}{4}$$
$$\sigma_T = \sqrt{\frac{n(n + 1)(2n + 1)}{24}}$$

##### Construindo a tabela com os postos

In [76]:
fumo

Antes  Depois
0      39      16
1      25       8
2      24      12
3      50       0
4      13      14
5      52      16
6      21      13
7      29      12
8      10      19
9      22      17
10     50      17
11     15       2
12     36      15
13     39      10
14     52      20
15     48      13
16     24       0
17     15       4
18     40      16
19     41      18
20     17      16
21     12      16
22     21       9
23     49       9
24     14      18
25     55       4
26     46      17
27     22       0
28     28      11
29     23      14
30     37       0
31     17      19
32     31       2
33     49       9
34     49       6

In [77]:
fumo['Diff'] = fumo.Depois - fumo.Antes

In [78]:
fumo.head()

Antes  Depois  Diff
0     39      16   -23
1     25       8   -17
2     24      12   -12
3     50       0   -50
4     13      14     1

In [79]:
fumo['|Diff|'] = fumo['Diff'].abs()

In [80]:
fumo.head()

Antes  Depois  Diff  |Diff|
0     39      16   -23      23
1     25       8   -17      17
2     24      12   -12      12
3     50       0   -50      50
4     13      14     1       1

In [81]:
fumo.sort_values(by='|Diff|', inplace=True)
fumo.head()

Antes  Depois  Diff  |Diff|
4      13      14     1       1
20     17      16    -1       1
31     17      19     2       2
21     12      16     4       4
24     14      18     4       4

In [82]:
fumo['Posto'] = range(1, len(fumo) + 1)
fumo.head()

Antes  Depois  Diff  |Diff|  Posto
4      13      14     1       1      1
20     17      16    -1       1      2
31     17      19     2       2      3
21     12      16     4       4      4
24     14      18     4       4      5

In [83]:
posto = fumo[['|Diff|', 'Posto']].groupby('|Diff|').mean()
posto

Posto
|Diff|       
1         1.5
2         3.0
4         4.5
5         6.0
8         7.0
9         8.5
11       10.0
12       11.5
13       13.0
17       15.0
21       17.0
22       18.0
23       19.5
24       21.5
29       24.0
32       26.0
33       27.0
35       28.0
36       29.0
37       30.0
40       31.5
43       33.0
50       34.0
51       35.0

In [84]:
posto.reset_index(inplace=True)
posto

|Diff|  Posto
0        1    1.5
1        2    3.0
2        4    4.5
3        5    6.0
4        8    7.0
5        9    8.5
6       11   10.0
7       12   11.5
8       13   13.0
9       17   15.0
10      21   17.0
11      22   18.0
12      23   19.5
13      24   21.5
14      29   24.0
15      32   26.0
16      33   27.0
17      35   28.0
18      36   29.0
19      37   30.0
20      40   31.5
21      43   33.0
22      50   34.0
23      51   35.0

In [85]:
fumo.drop('Posto', axis=1, inplace=True)
fumo.head()

Antes  Depois  Diff  |Diff|
4      13      14     1       1
20     17      16    -1       1
31     17      19     2       2
21     12      16     4       4
24     14      18     4       4

In [86]:
fumo = fumo.merge(posto, left_on='|Diff|', right_on='|Diff|', how='left')
fumo

Antes  Depois  Diff  |Diff|  Posto
0      13      14     1       1    1.5
1      17      16    -1       1    1.5
2      17      19     2       2    3.0
3      12      16     4       4    4.5
4      14      18     4       4    4.5
5      22      17    -5       5    6.0
6      21      13    -8       8    7.0
7      23      14    -9       9    8.5
8      10      19     9       9    8.5
9      15       4   -11      11   10.0
10     24      12   -12      12   11.5
11     21       9   -12      12   11.5
12     15       2   -13      13   13.0
13     28      11   -17      17   15.0
14     25       8   -17      17   15.0
15     29      12   -17      17   15.0
16     36      15   -21      21   17.0
17     22       0   -22      22   18.0
18     41      18   -23      23   19.5
19     39      16   -23      23   19.5
20     24       0   -24      24   21.5
21     40      16   -24      24   21.5
22     46      17   -29      29   24.0
23     39      10   -29      29   24.0
24     31       2   -29      29   24.0
25     52      20   -32      32   26.0
26     50      17   -33      33   27.0
27     48      13   -35      35   28.0
28     52      16   -36      36   29.0
29     37       0   -37      37   30.0
30     49       9   -40      40   31.5
31     49       9   -40      40   31.5
32     49       6   -43      43   33.0
33     50       0   -50      50   34.0
34     55       4   -51      51   35.0

In [87]:
fumo['Posto (+)'] = fumo.apply(lambda x: x.Posto if x.Diff > 0 else 0, axis=1)
fumo.head()

Antes  Depois  Diff  |Diff|  Posto  Posto (+)
0     13      14     1       1    1.5        1.5
1     17      16    -1       1    1.5        0.0
2     17      19     2       2    3.0        3.0
3     12      16     4       4    4.5        4.5
4     14      18     4       4    4.5        4.5

In [88]:
fumo['Posto (-)'] = fumo.apply(lambda x: x.Posto if x.Diff < 0 else 0, axis=1)
fumo.head()

Antes  Depois  Diff  |Diff|  Posto  Posto (+)  Posto (-)
0     13      14     1       1    1.5        1.5        0.0
1     17      16    -1       1    1.5        0.0        1.5
2     17      19     2       2    3.0        3.0        0.0
3     12      16     4       4    4.5        4.5        0.0
4     14      18     4       4    4.5        4.5        0.0

In [89]:
fumo.drop('Posto', axis=1, inplace=True)
fumo.head()

Antes  Depois  Diff  |Diff|  Posto (+)  Posto (-)
0     13      14     1       1        1.5        0.0
1     17      16    -1       1        0.0        1.5
2     17      19     2       2        3.0        0.0
3     12      16     4       4        4.5        0.0
4     14      18     4       4        4.5        0.0

##### Obter $T$

$T$ = menor das somas de postos de mesmo sinal

In [90]:
T = min(fumo['Posto (+)'].sum(), fumo['Posto (-)'].sum())
T

22.0

#### Obter $\mu_T$

$$\mu_T = \frac{n(n+1)}{4}$$

In [91]:
mu_T = (n * (n + 1)) / 4
mu_T

315.0

#### Obter $\sigma_T$

$$\sigma_T = \sqrt{\frac{n(n + 1)(2n + 1)}{24}}$$

In [92]:
sigma_T = np.sqrt((n * (n + 1) * ((2 * n) + 1)) / 24)
sigma_T

61.053255441458646

#### Obter $Z_{teste}$

$$Z = \frac{T - \mu_T}{\sigma_T}$$

In [93]:
Z = (T - mu_T) / sigma_T
Z

-4.799088891843698

![Estatística-Teste](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img021.png)

---

#### **Passo 5** - Aceitação ou rejeição da hipótese nula

<img src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img022.png' width='80%'>

#### <font color='red'>Critério do valor crítico</font>

> Rejeitar $H_0$ se $Z \leq -z_{\alpha / 2}$ ou se $Z \geq z_{\alpha / 2}$

In [94]:
Z <= z_alpha_2 or Z >= z_alpha_2

True

In [95]:
if (Z <= z_alpha_2 or Z >= z_alpha_2):
    print('Rejeita H0, ou seja, existe uma diferença entre os grupos.')
else:
    print('Aceita H0, ou seja, os grupos não apresentam diferença.')

Rejeita H0, ou seja, existe uma diferença entre os grupos.


<font color='green'>Conclusão: Rejeitamos a hipótese de que não existe diferença entre os grupos, isto é, existe uma diferença entre as médias de cigarros fumados pelos pacientes antes e depois do tratamento. E como é possível verificar através das médias de cigarros fumados por dia antes (31.86) e depois (11.2) do tratamento, podemos concluir que o tratamento apresentou resultado satisfatório.</font>

#### <font color='red'>Critério do valor $p$</font>

> Rejeitar $H_0$ se o valor $p\leq\alpha$

In [96]:
pvalue = 2 * norm.cdf(Z)

In [97]:
pvalue

1.5938904538674004e-06

In [98]:
pvalue <= significancia

True

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wilcoxon.html

In [122]:
from scipy.stats import wilcoxon

In [124]:
T, pvalue = wilcoxon(x=fumo.Antes, y=fumo.Depois)

In [125]:
if pvalue <= significancia:
    print('Rejeita H0')
else:
    print('Aceita H0')

Rejeita H0


---